### PR TITLE
[Serve] [Docs] Document how to build and deploy custom Docker images

### LINKED ISSUE #33457: [Serve] Documentation to build a custom docker image to deploy a serve application
### Description

There is no documentation about building a custom docker image to deploy a serve application. The current documentation only shows how to deploy it using remote zip files.

In a production environment, it is more common and a better workflow to build and deploy docker images rather than zipping up code. I did not think about building custom docker images until someone on the Ray slack mentioned that it could be done. Adding documentation to show this would make Serve more production friendly.

### Link

_No response_

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

In production, Serve users should put their Serve application code in a custom Docker image rather than a `runtime_env`. This change documents how to create custom Docker images and add them to a KubeRay config.

Link to docs: https://anyscale-ray--37370.com.readthedocs.build/en/37370/serve/production-guide/docker.html

## Related issue number

Closes #33457

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - The Serve app is tested in CI.

### CODE REVIEW

**Review by akshay-anyscale — APPROVED** [review]

File: doc/source/serve/production-guide/best-practices.md [hunk 1]
```diff
@@ -12,9 +12,10 @@ This section helps you:
...
-    * Put your deployment graph's code in a remote repository and manually configure the `working_dir` or `py_modules` fields in your Serve config file's `runtime_env` to point to that repository.
+* Use `serve run` to manually test and improve your Serve application locally.
+* Use `serve build` to create a Serve config file for your Serve application.
+    * For development, put your Serve application's code in a remote repository and manually configure the `working_dir` or `py_modules` fields in your Serve config file's `runtime_env` to point to that repository.
+    * For production, put your Serve application's code in a custom Docker image instead of a `runtime_env`. See [this tutorial](serve-custom-docker-images) to learn how to create custom Docker images and deploy them on KubeRay.
 * Use `serve status` to track your Serve application's health and deployment progress.
 * Use `serve config` to check the latest config that your Serve application received. This is its goal state.
 * Make lightweight configuration updates (e.g. `num_replicas` or `user_config` changes) by modifying your Serve config file and redeploying it with `serve deploy`.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/best-practices.md:21]:
```suggestion
* Make lightweight configuration updates (e.g., `num_replicas` or `user_config` changes) by modifying your Serve config file and redeploying it with `serve deploy`.
```

File: doc/source/serve/production-guide/docker.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# In another terminal window:
+% curl localhost:8000
+john24@example.org
+```
+
+This tutorial explains how to package and serve this code inside a custom Docker image.
+
+## Extending the Ray Docker image
+
+The [rayproject](https://hub.docker.com/u/rayproject) organization maintains a Docker images with dependencies needed to run Ray. In fact, the [rayproject/ray](https://hub.docker.com/r/rayproject/ray) and [rayproject/ray-ml](https://hub.docker.com/r/rayproject/ray-ml) repos host Docker images that we've used throughout the docs. For instance, [this RayService config](https://github.com/ray-project/kuberay/blob/release-0.6/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml) uses the [rayproject/ray:2.5.0](https://hub.docker.com/layers/rayproject/ray/2.5.0/images/sha256-cb53dcc21af8f913978fd2a3fc57c812f87d99e0b40db6a42ccd6f43eca11281) image hosted by `rayproject/ray`.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/docker.md:43]:
```suggestion
The [rayproject](https://hub.docker.com/u/rayproject) organization maintains a Docker images with dependencies needed to run Ray. In fact, the [rayproject/ray](https://hub.docker.com/r/rayproject/ray) and [rayproject/ray-ml](https://hub.docker.com/r/rayproject/ray-ml) repos host Docker images for this doc. For instance, [this RayService config](https://github.com/ray-project/kuberay/blob/release-0.6/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml) uses the [rayproject/ray:2.5.0](https://hub.docker.com/layers/rayproject/ray/2.5.0/images/sha256-cb53dcc21af8f913978fd2a3fc57c812f87d99e0b40db6a42ccd6f43eca11281) image hosted by `rayproject/ray`.
```

File: doc/source/serve/production-guide/docker.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+You can extend these images and add your own dependencies to them by using them as a base layer in a Dockerfile. For instance, the working example application uses Ray 2.5.0 and Faker 18.13.0. You can create a Dockerfile that extends the `rayproject/ray:2.5.0` by adding the Faker package:
+
+```dockerfile
+# File name: Dockerfile
+FROM rayproject/ray:2.5.0
+
+RUN pip install Faker==18.13.0
+```
+
+In general, the `rayproject/ray` images contain only the dependencies needed to import Ray and the Ray libraries. The `rayproject/ray-ml` images contain additional dependencies (e.g. PyTorch, HuggingFace, etc.) that are useful for machine learning. You can extend images from either of these repos to build your custom images.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/docker.md:54]:
```suggestion
In general, the `rayproject/ray` images contain only the dependencies needed to import Ray and the Ray libraries. The `rayproject/ray-ml` images contain additional dependencies (e.g., PyTorch, HuggingFace, etc.) that are useful for machine learning. You can extend images from either of these repos to build your custom images.
```

**Review by angelinalg — APPROVED** [review]
Just had some minor nits. Nicely written!

**Comment by shrekris-anyscale** [comment]
> Just had some minor nits. Nicely written!

Thanks for reviewing!

**Comment by shrekris-anyscale** [comment]
The test failures are unrelated:

* Windows Build & Test ([`test_object_store_metrics`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/28719#018965ca-13f1-44f0-8cc1-514db46f1dac/872-1266)) is not affected by docs changes.
* [`Linkcheck`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/28719#018965d6-f9b3-4cb5-a973-9640e07621e7/1206-5795) is failing due to links not added by this change:

```console
BROKEN LINKS SUMMARY:
--
  |  
  | (rllib/rllib-examples: line  218) broken    https://github.com/awslabs/amazon-sagemaker-examples/tree/master/reinforcement_learning/rl_roboschool_ray - 404 Client Error: Not Found for url: https://github.com/aws/amazon-sagemaker-examples/tree/main/reinforcement_learning/rl_roboschool_ray
  | (ray-overview/examples: line  695) broken    https://www.uber.com/blog/elastic-xgboost-ray/ - 406 Client Error: Not Acceptable for url: https://www.uber.com/blog/elastic-xgboost-ray/
  | make: *** [Makefile:181: linkcheck] Error 1
  | 🚨 Error: The command exited with status 2
```

**Comment by shrekris-anyscale** [comment]
@edoakes @architkulkarni This change is ready to merge.

**Pull request merged**